Add Pulsar configuration to determine the collection of message contents

CommissarXia opened the pull request:
<!--
    ⚠️ Please make sure to read this template first, pull requests that don't accord with this template
    maybe closed without notice.
    Texts surrounded by `<` and `>` are meant to be replaced by you, e.g. <framework name>, <issue number>.
    Put an `x` in the `[ ]` to mark the item as CHECKED. `[x]`
-->

<!-- ==== 🐛 Remove this line WHEN AND ONLY WHEN you're fixing a bug, follow the checklist 👇 ====
### Fix <bug description or the bug issue number or bug issue link>
- [ ] Add a unit test to verify that the fix works.
- [ ] Explain briefly why the bug exists and how to fix it.
     ==== 🐛 Remove this line WHEN AND ONLY WHEN you're fixing a bug, follow the checklist 👆 ==== -->

<!-- ==== 🔌 Remove this line WHEN AND ONLY WHEN you're adding a new plugin, follow the checklist 👇 ====
### Add an agent plugin to support <framework name>
- [ ] Add a test case for the new plugin, refer to [the doc](https://github.com/apache/skywalking-java/blob/main/docs/en/setup/service-agent/java-agent/Plugin-test.md)
- [ ] Add a component id in [the component-libraries.yml](https://github.com/apache/skywalking/blob/master/oap-server/server-starter/src/main/resources/component-libraries.yml)
- [ ] Add a logo in [the UI repo](https://github.com/apache/skywalking-rocketbot-ui/tree/master/src/views/components/topology/assets)
     ==== 🔌 Remove this line WHEN AND ONLY WHEN you're adding a new plugin, follow the checklist 👆 ==== -->

<!-- ==== 📈 Remove this line WHEN AND ONLY WHEN you're improving the performance, follow the checklist 👇 ====
### Improve the performance of <class or module or ...>
- [ ] Add a benchmark for the improvement, refer to [the existing ones](https://github.com/apache/skywalking-java/blob/main/apm-commons/apm-datacarrier/src/test/java/org/apache/skywalking/apm/commons/datacarrier/LinkedArrayBenchmark.java)
- [ ] The benchmark result.
```text
<Paste the benchmark results here>
```
- [ ] Links/URLs to the theory proof or discussion articles/blogs. <links/URLs here>
     ==== 📈 Remove this line WHEN AND ONLY WHEN you're improving the performance, follow the checklist 👆 ==== -->

### <Feature description>
- [ ] If this is non-trivial feature, paste the links/URLs to the design doc.
- [ ] Update the documentation to include this new feature.
- [x] Tests(including UT, IT, E2E) are added to verify the new feature.
- [ ] If it's UI related, attach the screenshots below.
- [x] Modify the unit test to verify that the feature works.
- [x] The configuration has been verified in practice.

- [ ] If this pull request closes/resolves/fixes an existing issue, replace the issue number. Closes #<issue number>.
- [ ] Update the [`CHANGES` log](https://github.com/apache/skywalking-java/blob/main/CHANGES.md).

wu-sheng commented on apm-sniffer/apm-sdk-plugin/pulsar-common/src/main/java/org/apache/skywalking/apm/plugin/pulsar/common/PulsarPluginConfig.java:
```diff
@@ -0,0 +1,42 @@
+            /**
+             * If set to true, the message contents of the Pulsar topic would be collected.
+             */
+            public static boolean TRACE_MESSAGE_CONTENTS = true;
```
The default value should be false. The impact is clear for memory, network and database.

wu-sheng replied:
agent.config set to false is not enough.

CommissarXia replied:
Okay, I'll make some modifications and conduct unit testing. Thank you!

wu-sheng commented on apm-sniffer/apm-sdk-plugin/pulsar-common/src/main/java/org/apache/skywalking/apm/plugin/pulsar/common/PulsarConsumerInterceptor.java:
```diff
@@ -73,6 +75,9 @@ public void beforeMethod(EnhancedInstance objInst, Method method, Object[] allAr
             Tags.MQ_TOPIC.set(activeSpan, consumer.getTopic());
             activeSpan.setPeer(serviceUrl);
+            if (Pulsar.TRACE_MESSAGE_CONTENTS) {
+                Tags.MQ_BODY.set(activeSpan, StringUtil.cut(new String(msg.getData()), Pulsar.MESSAGE_CONTENTS_MAX_LENGTH));
```
`msg.getData()` is a byte[]. Is here OK to use the default charset?

CommissarXia replied:
We used the default encoding for a long time without any issues, but we are using Java and are unsure if other languages can execute properly. This time, I added the UTF-8 character set and it ran normally in our environment

wu-sheng commented:
Please update this test, https://github.com/apache/skywalking-java/tree/main/test/plugin/scenarios/pulsar-scenario

Plugin test could run locally, https://skywalking.apache.org/docs/skywalking-java/v9.1.0/en/setup/service-agent/java-agent/plugin-test/

You should activate this new config and verify new tags.

CommissarXia commented:
There are some issues with the environment of my computer. I will submit it after I run it. Thanks!

wu-sheng commented on test/plugin/scenarios/pulsar-scenario/config/expectedData.yaml:
```diff
@@ -97,6 +99,7 @@ segmentItems:
             - { key: transmission.latency, value: not blank }
             - { key: mq.broker, value: not blank }
             - { key: mq.topic, value: test }
+            - { key: mq.body, value: "1" }
```
This is how the message is built, 
```java
producer.newMessage().key("testKey").value(Integer.toString(1).getBytes()).property("TEST", "TEST").send();
```

But you only set `1` as body? Could you explain?

CommissarXia replied:
This 1 was set by the previous submitter, and I only tested the corresponding expected value based on this value. If need test message bodies in other formats, I can modify and resubmit them.

wu-sheng replied:
But there is not only value. There are a key and properties. aren't those message body too?

wu-sheng commented:
Could you explain what you changed? 

Personally, I am frustrated about the communications between us. I have to read codes, understand codes, but you never relies words.

CommissarXia commented:
Actually, I only added a dynamic configuration item to dynamically collect the message contents of Pulsar. I changed "mq.body" to "mq.message_contents" and have already submitted it. This can more clearly reflect the content I added and the expected results.

wu-sheng commented:
My question is key/value, why is the key not collected?

CommissarXia commented:
The key is Pulsar's strategy for message routing, and its relationship with actual messages can be inclusive or independent. I just wanted to collect actual messages, so I didn't consider this key. Perhaps my consideration was a bit lacking.

wu-sheng commented:
Is there pulsar doc indicating key is not a part of message? I am doubt about that.

wu-sheng commented:
@tisonkun What is the message concept in the Pulsar?

CommissarXia commented:
Misunderstand my meaning, key must be passed together with the content of the message body, but the collection content I want to add does not take into account key, I will add it sometime today.

CommissarXia commented:
The messages in Pulsar include Value/data payload, Key, Properties, etc. My initial idea was just to collect Values.

wu-sheng commented:
That doesn't make sense from my understanding, unless this is from Pulsar officially.
If we declare collecting message, it should be the whole.

CommissarXia commented:
OK, it should be my lack of understanding. I'll readjust it again. I'm really sorry for wasting your time.

wu-sheng commented:
For next time, please write the proposal that you design first. It would help me align with you.

CommissarXia commented:
Ok, thank you very much!

wu-sheng commented:
Any update about this/

CommissarXia commented:
I was sick last week. I have time to deal with it this week.

wu-sheng commented:
Is this still on going?

CommissarXia commented:
Shengge, After careful consideration later, maybe this feature is not very necessary, and the demand for simply collecting message content is probably not too much, so I want to give up this pull request. 
Shengge, do you have any opinions? Sorry for wasting your time.

wu-sheng commented:
Closing upon request.
